include/sparrow/allocator.hpp

```diff
@@ -0,0 +1,250 @@
+
+        any_allocator& operator=(const any_allocator& rhs) = delete;
+        any_allocator& operator=(any_allocator&& rhs) = delete;
+
+        template <class A>
+        any_allocator(A&& alloc)
+        requires (not std::same_as<std::remove_cvref_t<A>, any_allocator>
+                and sparrow::allocator<A>)
+            : m_storage(make_storage(std::forward<A>(alloc)))
+        {
+        }
+
+        [[nodiscard]] T* allocate(std::size_t n);
+        void deallocate(T* p, std::size_t n);
+        
+        any_allocator select_on_container_copy_construction() const;
+
+        bool equal(const any_allocator& rhs) const; 
+
+    private:
+
+        struct interface
+        {
+            [[nodiscard]] virtual T* allocate(std::size_t) = 0;
+            virtual void deallocate(T*, std::size_t) = 0;
+            virtual std::unique_ptr<interface> clone() const = 0;
+            virtual bool equal(const interface&) const = 0;
+            virtual ~interface() = default;
+        };
+
+        template <allocator A>
+        struct impl : interface
+        {
+            A m_alloc;
+
+            explicit impl(A alloc) : m_alloc(std::move(alloc)) {}
+
+            [[nodiscard]] T* allocate(std::size_t n) override
+            {
+                return m_alloc.allocate(n);
+            }
+
+            void deallocate(T* p, std::size_t n) override
+            {
+                m_alloc.deallocate(p, n);
+            }
+
+            std::unique_ptr<interface> clone() const override
+            {
+                return std::make_unique<impl<A>>(m_alloc);
+            }
+
+            bool equal(const interface& rhs) const override
+            {
+                if (std::type_index(typeid(*this)) == std::type_index(typeid(rhs)))
+                {
+                    return m_alloc == static_cast<const impl<A>*>(std::addressof(rhs))->m_alloc;
+                }
+                return false;
+            }
+        };
+
+        using storage_type = std::variant
+        <
+            std::allocator<T>,
+            std::pmr::polymorphic_allocator<T>,
+            std::unique_ptr<interface>
+        >;
+
+        template <class A>
+        std::unique_ptr<interface> make_storage(A&& alloc) const
+        {
+            return std::make_unique<impl<std::decay_t<A>>>(std::forward<A>(alloc));
+        }
+
+        template <class A>
+        requires can_any_allocator_sbo<A, T>
+        A&& make_storage(A&& alloc) const
+        {
+            return std::forward<A>(alloc);
+        }
+
+        template <class... Ts>
+        struct overloaded : Ts... { using Ts::operator()...; };
+
+        storage_type copy_storage(const storage_type& rhs) const
+        {
+            return std::visit(overloaded {
+                [](const auto& arg) -> storage_type { return { std::decay_t<decltype(arg)>(arg) }; },
+                [](const std::unique_ptr<interface>& arg) -> storage_type { return { arg->clone() }; }
+            }, rhs);
```

Comment:
Change done, I haven't reuse `visit_storage` in the implementation of `copy_storage` for the following reasons:
- copy_storage is const, while allocate and deallocate are not
- copy_storage operates on another storage than `m_storage`

=> This means we need a second overload of the `visit_storage` method to have the second lambda accept `const interface&` instead of `const std::unique_ptr<interface>&`. I'm not sure adding so much code for changing an argument is worth it  (i.e. I find the code understandable without it), but happy to do it if you think so.

Reply:
I feel it's fine like this